File: tools/import_asset.py
#!/usr/bin/env python3
"""Deterministic, game-independent asset import primitives."""
import argparse
import json
import struct
import zlib
from pathlib import Path


def chunks(data, signature):
    if not data.startswith(signature):
        raise ValueError("invalid file signature")
    cursor = len(signature)
    while cursor < len(data):
        if cursor + 12 > len(data):
            raise ValueError("truncated chunk")
        size = struct.unpack_from(">I", data, cursor)[0]
        kind = data[cursor + 4:cursor + 8]
        end = cursor + 12 + size
        if end > len(data):
            raise ValueError("truncated chunk payload")
        payload = data[cursor + 8:cursor + 8 + size]
        expected = struct.unpack_from(">I", data, cursor + 8 + size)[0]
        if zlib.crc32(kind + payload) & 0xffffffff != expected:
            raise ValueError("chunk CRC mismatch")
        yield kind, payload
        cursor = end


def paeth(left, above, upper_left):
    prediction = left + above - upper_left
    choices = (abs(prediction - left), abs(prediction - above),
               abs(prediction - upper_left))
    return (left, above, upper_left)[choices.index(min(choices))]


def indexed_png(data):
    width = height = depth = None
    palette = transparency = None
    compressed = bytearray()
    for kind, payload in chunks(data, b"\x89PNG\r\n\x1a\n"):
        if kind == b"IHDR":
            if len(payload) != 13:
                raise ValueError("invalid IHDR")
            width, height, depth, color, compression, filtering, interlace = (
                struct.unpack(">IIBBBBB", payload))
            if color != 3 or depth not in (1, 2, 4, 8):
                raise ValueError("PNG must use an indexed 1/2/4/8-bit palette")
            if compression or filtering or interlace:
                raise ValueError("unsupported PNG encoding")
        elif kind == b"PLTE":
            if len(payload) % 3 or not 3 <= len(payload) <= 768:
                raise ValueError("invalid PLTE")
            palette = [tuple(payload[i:i + 3])
                       for i in range(0, len(payload), 3)]
        elif kind == b"tRNS":
            transparency = payload
        elif kind == b"IDAT":
            compressed.extend(payload)
        elif kind == b"IEND":
            break
    if None in (width, height, depth) or palette is None:
        raise ValueError("PNG lacks IHDR or PLTE")
    if width == 0 or height == 0 or width % 8 or height % 8:
        raise ValueError("PNG dimensions must be nonzero multiples of eight")
    if len(palette) > 1 << depth:
        raise ValueError("palette exceeds indexed bit depth")
    if transparency is not None and (
            len(transparency) > len(palette) or
            any(alpha not in (0, 255) for alpha in transparency)):
        raise ValueError("transparency must contain only binary alpha")
    row_size = (width * depth + 7) // 8
    raw = zlib.decompress(bytes(compressed))
    if len(raw) != height * (row_size + 1):
        raise ValueError("unexpected decompressed PNG size")
    rows = []
    cursor = 0
    previous = bytearray(row_size)
    for _ in range(height):
        method = raw[cursor]
        source = raw[cursor + 1:cursor + 1 + row_size]
        cursor += row_size + 1
        if method > 4:
            raise ValueError("invalid PNG filter")
        row = bytearray(row_size)
        for index, value in enumerate(source):
            left = row[index - 1] if index else 0
            above = previous[index]
            upper_left = previous[index - 1] if index else 0
            predictor = (0, left, above, (left + above) // 2,
                         paeth(left, above, upper_left))[method]
            row[index] = (value + predictor) & 255
        rows.append(row)
        previous = row
    pixels = []
    mask = (1 << depth) - 1
    for row in rows:
        for x in range(width):
            shift = 8 - depth - (x * depth % 8)
            index = row[x * depth // 8] >> shift & mask
            if index >= len(palette):
                raise ValueError("pixel references a missing palette entry")
            pixels.append(index)
    return width, height, pixels, palette


def rgba_png(data):
    width = height = None
    compressed = bytearray()
    for kind, payload in chunks(data, b"\x89PNG\r\n\x1a\n"):
        if kind == b"IHDR":
            if len(payload) != 13:
                raise ValueError("invalid IHDR")
            width, height, depth, color, compression, filtering, interlace = (
                struct.unpack(">IIBBBBB", payload))
            if color != 6 or depth != 8:
                raise ValueError("PNG must use 8-bit RGBA pixels")
            if compression or filtering or interlace:
                raise ValueError("unsupported PNG encoding")
        elif kind == b"IDAT":
            compressed.extend(payload)
        elif kind == b"IEND":
            break
    if width is None or height is None or width == 0 or height == 0:
        raise ValueError("PNG lacks a nonempty IHDR")
    row_size = width * 4
    raw = zlib.decompress(bytes(compressed))
    if len(raw) != height * (row_size + 1):
        raise ValueError("unexpected decompressed PNG size")
    rows = []
    cursor = 0
    previous = bytearray(row_size)
    for _ in range(height):
        method = raw[cursor]
        source = raw[cursor + 1:cursor + row_size + 1]
        cursor += row_size + 1
        if method > 4:
            raise ValueError("invalid PNG filter")
        row = bytearray(row_size)
        for index, value in enumerate(source):
            left = row[index - 4] if index >= 4 else 0
            above = previous[index]
            upper_left = previous[index - 4] if index >= 4 else 0
            predictor = (0, left, above, (left + above) // 2,
                         paeth(left, above, upper_left))[method]
            row[index] = (value + predictor) & 255
        rows.append(row)
        previous = row
    return width, height, b"".join(rows)


def gba_palette_rgba(data):
    width, height, pixels = rgba_png(data)
    palette = bytearray()
    for offset in range(0, len(pixels), 4):
        red, green, blue, alpha = pixels[offset:offset + 4]
        if red & 7 or green & 7 or blue & 7:
            raise ValueError("RGBA palette colors must be multiples of eight")
        if alpha not in (254, 255):
            raise ValueError("RGBA palette alpha must be 254 or 255")
        value = ((red >> 3) | (green >> 3) << 5 |
                 (blue >> 3) << 10 | (255 - alpha) << 15)
        palette.extend(struct.pack("<H", value))
    return bytes(palette), {"width": width, "height": height,
                            "palette_entries": len(palette) // 2}


def gba_graphics(data, bpp):
    width, height, pixels, palette = indexed_png(data)
    limit = 16 if bpp == 4 else 256
    if len(palette) > limit or any(pixel >= limit for pixel in pixels):
        raise ValueError(f"image does not fit {bpp}bpp")
    colors = bytearray()
    for red, green, blue in palette:
        if red & 7 or green & 7 or blue & 7:
            raise ValueError(
                "palette channels must be exact five-bit values (multiples of 8)")
        value = (red >> 3) | (green >> 3) << 5 | (blue >> 3) << 10
        colors.extend(struct.pack("<H", value))
    tiles = bytearray()
    for top in range(0, height, 8):
        for left in range(0, width, 8):
            tile = [pixels[(top + y) * width + left + x]
                    for y in range(8) for x in range(8)]
            if bpp == 4:
                tiles.extend(tile[i] | tile[i + 1] << 4
                             for i in range(0, 64, 2))
            else:
                tiles.extend(tile)
    return bytes(tiles), bytes(colors), {
        "width": width, "height": height, "bpp": bpp,
        "tiles": width // 8 * (height // 8),
        "palette_entries": len(palette),
    }


def vlq(data, cursor):
    value = 0
    for _ in range(4):
        if cursor >= len(data):
            raise ValueError("truncated variable-length quantity")
        byte = data[cursor]
        cursor += 1
        value = value << 7 | byte & 0x7f
        if byte < 0x80:
            return value, cursor
    raise ValueError("overlong variable-length quantity")


def midi_events(data):
    parsed = []
    cursor = 0
    while cursor < len(data):
        if cursor + 8 > len(data):
            raise ValueError("truncated MIDI chunk")
        kind = data[cursor:cursor + 4]
        size = struct.unpack_from(">I", data, cursor + 4)[0]
        cursor += 8
        payload = data[cursor:cursor + size]
        cursor += size
        if len(payload) != size:
            raise ValueError("truncated MIDI chunk payload")
        parsed.append((kind, payload))
    if not parsed or parsed[0][0] != b"MThd" or len(parsed[0][1]) != 6:
        raise ValueError("invalid MIDI header")
    form, tracks, division = struct.unpack(">HHH", parsed[0][1])
    if form not in (0, 1) or tracks == 0 or division & 0x8000:
        raise ValueError("only format 0/1 PPQN MIDI is supported")
    track_chunks = [payload for kind, payload in parsed[1:] if kind == b"MTrk"]
    if len(track_chunks) != tracks:
        raise ValueError("MIDI track count mismatch")
    result = []
    lengths = {0xC0: 1, 0xD0: 1}
    for track, payload in enumerate(track_chunks):
        cursor = tick = order = 0
        running = None
        while cursor < len(payload):
            delta, cursor = vlq(payload, cursor)
            tick += delta
            if cursor >= len(payload):
                raise ValueError("truncated MIDI event")
            status = payload[cursor]
            if status < 0x80:
                if running is None:
                    raise ValueError("running status without channel status")
                status = running
            else:
                cursor += 1
            event = {"tick": tick, "track": track, "order": order}
            order += 1
            if status == 0xff:
                if cursor >= len(payload):
                    raise ValueError("truncated meta event")
                kind = payload[cursor]
                size, cursor = vlq(payload, cursor + 1)
                value = payload[cursor:cursor + size]
                cursor += size
                if len(value) != size:
                    raise ValueError("truncated meta payload")
                event.update(type="meta", meta=kind, data=value.hex())
                running = None
            elif status in (0xf0, 0xf7):
                size, cursor = vlq(payload, cursor)
                value = payload[cursor:cursor + size]
                cursor += size
                if len(value) != size:
                    raise ValueError("truncated system-exclusive payload")
                event.update(type="sysex", status=status, data=value.hex())
                running = None
            elif 0x80 <= status <= 0xef:
                running = status
                size = lengths.get(status & 0xf0, 2)
                value = payload[cursor:cursor + size]
                cursor += size
                if len(value) != size or any(byte & 0x80 for byte in value):
                    raise ValueError("invalid channel event")
                event.update(type="channel", status=status,
                             data=list(value))
            else:
                raise ValueError("unsupported MIDI system event")
            result.append(event)
    result.sort(key=lambda item: (item["tick"], item["track"], item["order"]))
    return {"format": form, "tracks": tracks, "ticks_per_quarter": division,
            "events": result}


def make_chunk(kind, payload):
    return (struct.pack(">I", len(payload)) + kind + payload +
            struct.pack(">I", zlib.crc32(kind + payload) & 0xffffffff))


def make_midi_chunk(kind, payload):
    return kind + struct.pack(">I", len(payload)) + payload


def self_test():
    palette = bytes((0, 0, 0, 248, 0, 0))
    rows = b"".join(b"\0" + bytes((0x55,)) for _ in range(8))
    png = (b"\x89PNG\r\n\x1a\n" +
           make_chunk(b"IHDR", struct.pack(">IIBBBBB", 8, 8, 1, 3, 0, 0, 0)) +
           make_chunk(b"PLTE", palette) +
           make_chunk(b"IDAT", zlib.compress(rows)) +
           make_chunk(b"IEND", b""))
    tiles, colors, report = gba_graphics(png, 4)
    assert tiles == bytes((0x10,)) * 32
    assert colors == b"\0\0\x1f\0" and report["tiles"] == 1
    track = b"\0\xc0\x05\0\x90\x3c\x40\x60\x80\x3c\0\0\xff\x2f\0"
    midi = (make_midi_chunk(b"MThd", struct.pack(">HHH", 0, 1, 96)) +
            make_midi_chunk(b"MTrk", track))
    report = midi_events(midi)
    assert report["ticks_per_quarter"] == 96
    assert [event["tick"] for event in report["events"]] == [0, 0, 96, 96]
    rgba = bytes((index * 29 + 3) & 255 for index in range(7 * 5 * 4))
    from export_asset import rgba_image
    image, _ = rgba_image(rgba, 7)
    assert rgba_png(image) == (7, 5, rgba)
    raw_palette = b"".join(struct.pack("<H", (index * 109) & 0xffff)
                           for index in range(128))
    from export_asset import palette_rgba_image
    image, _ = palette_rgba_image(raw_palette, 16)
    assert gba_palette_rgba(image)[0] == raw_palette
    print("self-test=ok")


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("--self-test", action="store_true")
    commands = parser.add_subparsers(dest="command")
    png = commands.add_parser("png")
    png.add_argument("input", type=Path)
    png.add_argument("--bpp", type=int, choices=(4, 8), required=True)
    png.add_argument("--tiles", type=Path, required=True)
    png.add_argument("--palette", type=Path, required=True)
    indexed = commands.add_parser("indexed")
    indexed.add_argument("input", type=Path)
    indexed.add_argument("-o", "--output", type=Path, required=True)
    midi = commands.add_parser("midi")
    midi.add_argument("input", type=Path)
    midi.add_argument("-o", "--output", type=Path, required=True)
    args = parser.parse_args()
    if args.self_test:
        self_test()
        if args.command is None:
            return
    if args.command == "png":
        tiles, palette, report = gba_graphics(args.input.read_bytes(), args.bpp)
        args.tiles.parent.mkdir(parents=True, exist_ok=True)
        args.palette.parent.mkdir(parents=True, exist_ok=True)
        args.tiles.write_bytes(tiles)
        args.palette.write_bytes(palette)
        print(json.dumps(report, sort_keys=True))
    elif args.command == "indexed":
        width, height, pixels, palette = indexed_png(args.input.read_bytes())
        args.output.parent.mkdir(parents=True, exist_ok=True)
        args.output.write_bytes(bytes(pixels))
        print(json.dumps({"width": width, "height": height,
                          "pixels": len(pixels),
                          "palette_entries": len(palette)}, sort_keys=True))
    elif args.command == "midi":
        report = midi_events(args.input.read_bytes())
        args.output.parent.mkdir(parents=True, exist_ok=True)
        args.output.write_text(json.dumps(report, indent=2) + "\n")
        print(f"tracks={report['tracks']} events={len(report['events'])}")
    else:
        parser.error("an asset command is required")


if __name__ == "__main__":
    main()

File: tools/import_asset.ts
// Tool role: both; imported by tools/archive_asset.ts, tools/battle_effect_data.ts, tools/build_assets.ts (+27 more); invoked by CONVENTIONS.md, package.json.
import { mkdirSync, readFileSync, writeFileSync } from "node:fs";
import { dirname } from "node:path";
import { crc32, deflateSync, inflateSync } from "./zlib.ts";
import { canonicalJson } from "./canonical_json.ts";

export type Rgb = [number, number, number];
export type Report = Record<string, number>;

export function* chunks(data: Uint8Array, signature: Uint8Array): Generator<[Buffer, Buffer]> {
  const source = Buffer.from(data.buffer, data.byteOffset, data.byteLength);
  const marker = Buffer.from(signature.buffer, signature.byteOffset, signature.byteLength);
  if (!source.subarray(0, marker.length).equals(marker)) throw new Error("invalid file signature");
  let cursor = marker.length;
  while (cursor < source.length) {
    if (cursor + 12 > source.length) throw new Error("truncated chunk");
    const size = source.readUInt32BE(cursor);
    const kind = source.subarray(cursor + 4, cursor + 8);
    const end = cursor + 12 + size;
    if (end > source.length) throw new Error("truncated chunk payload");
    const payload = source.subarray(cursor + 8, cursor + 8 + size);
    const expected = source.readUInt32BE(cursor + 8 + size);
    if (crc32(Buffer.concat([kind, payload])) !== expected) throw new Error("chunk CRC mismatch");
    yield [kind, payload];
    cursor = end;
  }
}

export function paeth(left: number, above: number, upper_left: number): number {
  const prediction = left + above - upper_left;
  const a = Math.abs(prediction - left);
  const b = Math.abs(prediction - above);
  const c = Math.abs(prediction - upper_left);
  return a <= b && a <= c ? left : b <= c ? above : upper_left;
}

function unfilter(raw: Uint8Array, height: number, row_size: number, stride: number): Buffer[] {
  if (raw.length !== height * (row_size + 1)) throw new Error("unexpected decompressed PNG size");
  const rows: Buffer[] = [];
  let cursor = 0;
  let previous = Buffer.alloc(row_size);
  for (let y = 0; y < height; y++) {
    const method = raw[cursor];
    const source = raw.subarray(cursor + 1, cursor + 1 + row_size);
    cursor += row_size + 1;
    if (method > 4) throw new Error("invalid PNG filter");
    const row = Buffer.alloc(row_size);
    for (let index = 0; index < row_size; index++) {
      const left = index >= stride ? row[index - stride] : 0;
      const above = previous[index];
      const upperLeft = index >= stride ? previous[index - stride] : 0;
      const predictor = method === 0 ? 0 : method === 1 ? left : method === 2 ? above :
        method === 3 ? Math.floor((left + above) / 2) : paeth(left, above, upperLeft);
      row[index] = (source[index] + predictor) & 255;
    }
    rows.push(row);
    previous = row;
  }
  return rows;
}

export function indexed_png(data: Uint8Array): [number, number, number[], Rgb[]] {
  let width: number | undefined;
  let height: number | undefined;
  let depth: number | undefined;
  let palette: Rgb[] | undefined;
  let transparency: Buffer | undefined;
  const compressed: Buffer[] = [];
  for (const [kind, payload] of chunks(data, Buffer.from("89504e470d0a1a0a", "hex"))) {
    const name = kind.toString("ascii");
    if (name === "IHDR") {
      if (payload.length !== 13) throw new Error("invalid IHDR");
      width = payload.readUInt32BE(0);
      height = payload.readUInt32BE(4);
      depth = payload[8];
      const color = payload[9];
      if (color !== 3 || ![1, 2, 4, 8].includes(depth))
        throw new Error("PNG must use an indexed 1/2/4/8-bit palette");
      if (payload[10] || payload[11] || payload[12]) throw new Error("unsupported PNG encoding");
    } else if (name === "PLTE") {
      if (payload.length % 3 !== 0 || payload.length < 3 || payload.length > 768) throw new Error("invalid PLTE");
      palette = [];
      for (let index = 0; index < payload.length; index += 3)
        palette.push([payload[index], payload[index + 1], payload[index + 2]]);
    } else if (name === "tRNS") transparency = Buffer.from(payload);
    else if (name === "IDAT") compressed.push(payload);
    else if (name === "IEND") break;
  }
  if (width === undefined || height === undefined || depth === undefined || palette === undefined)
    throw new Error("PNG lacks IHDR or PLTE");
  if (width === 0 || height === 0 || width % 8 !== 0 || height % 8 !== 0)
    throw new Error("PNG dimensions must be nonzero multiples of eight");
  if (palette.length > 1 << depth) throw new Error("palette exceeds indexed bit depth");
  if (transparency && (transparency.length > palette.length || Array.from(transparency).some((alpha) => alpha !== 0 && alpha !== 255)))
    throw new Error("transparency must contain only binary alpha");
  const rowSize = Math.floor((width * depth + 7) / 8);
  const rows = unfilter(inflateSync(Buffer.concat(compressed)), height, rowSize, 1);
  const pixels: number[] = [];
  const mask = (1 << depth) - 1;
  for (const row of rows) for (let x = 0; x < width; x++) {
    const shift = 8 - depth - (x * depth % 8);
    const index = row[Math.floor(x * depth / 8)] >>> shift & mask;
    if (index >= palette.length) throw new Error("pixel references a missing palette entry");
    pixels.push(index);
  }
  return [width, height, pixels, palette];
}

export function rgba_png(data: Uint8Array): [number, number, Buffer] {
  let width: number | undefined;
  let height: number | undefined;
  const compressed: Buffer[] = [];
  for (const [kind, payload] of chunks(data, Buffer.from("89504e470d0a1a0a", "hex"))) {
    const name = kind.toString("ascii");
    if (name === "IHDR") {
      if (payload.length !== 13) throw new Error("invalid IHDR");
      width = payload.readUInt32BE(0);
      height = payload.readUInt32BE(4);
      if (payload[9] !== 6 || payload[8] !== 8) throw new Error("PNG must use 8-bit RGBA pixels");
      if (payload[10] || payload[11] || payload[12]) throw new Error("unsupported PNG encoding");
    } else if (name === "IDAT") compressed.push(payload);
    else if (name === "IEND") break;
  }
  if (width === undefined || height === undefined || width === 0 || height === 0)
    throw new Error("PNG lacks a nonempty IHDR");
  const rows = unfilter(inflateSync(Buffer.concat(compressed)), height, width * 4, 4);
  return [width, height, Buffer.concat(rows)];
}

export function gba_palette_rgba(data: Uint8Array): [Buffer, Report] {
  const [width, height, pixels] = rgba_png(data);
  const palette = Buffer.alloc(pixels.length / 2);
  for (let offset = 0; offset < pixels.length; offset += 4) {
    const [red, green, blue, alpha] = pixels.subarray(offset, offset + 4);
    if ((red & 7) || (green & 7) || (blue & 7)) throw new Error("RGBA palette colors must be multiples of eight");
    if (alpha !== 254 && alpha !== 255) throw new Error("RGBA palette alpha must be 254 or 255");
    palette.writeUInt16LE((red >>> 3) | (green >>> 3) << 5 | (blue >>> 3) << 10 | (255 - alpha) << 15, offset / 2);
  }
  return [palette, { width, height, palette_entries: palette.length / 2 }];
}

export function gba_graphics(data: Uint8Array, bpp: number): [Buffer, Buffer, Report] {
  const [width, height, pixels, palette] = indexed_png(data);
  const limit = bpp === 4 ? 16 : 256;
  if (palette.length > limit || pixels.some((pixel) => pixel >= limit)) throw new Error(`image does not fit ${bpp}bpp`);
  const colors = Buffer.alloc(palette.length * 2);
  palette.forEach(([red, green, blue], index) => {
    if ((red & 7) || (green & 7) || (blue & 7))
      throw new Error("palette channels must be exact five-bit values (multiples of 8)");
    colors.writeUInt16LE((red >>> 3) | (green >>> 3) << 5 | (blue >>> 3) << 10, index * 2);
  });
  const tiles: number[] = [];
  for (let top = 0; top < height; top += 8) for (let left = 0; left < width; left += 8) {
    const tile: number[] = [];
    for (let y = 0; y < 8; y++) for (let x = 0; x < 8; x++) tile.push(pixels[(top + y) * width + left + x]);
    if (bpp === 4) for (let index = 0; index < 64; index += 2) tiles.push(tile[index] | tile[index + 1] << 4);
    else tiles.push(...tile);
  }
  return [Buffer.from(tiles), colors, { width, height, bpp, tiles: width / 8 * (height / 8), palette_entries: palette.length }];
}

export function vlq(data: Uint8Array, cursor: number): [number, number] {
  let value = 0;
  for (let index = 0; index < 4; index++) {
    if (cursor >= data.length) throw new Error("truncated variable-length quantity");
    const byte = data[cursor++];
    value = value << 7 | byte & 0x7f;
    if (byte < 0x80) return [value, cursor];
  }
  throw new Error("overlong variable-length quantity");
}

export interface MidiEvent { tick: number; track: number; order: number; type?: string; meta?: number; status?: number; data?: string | number[] }
export interface MidiReport { format: number; tracks: number; ticks_per_quarter: number; events: MidiEvent[] }

export function midi_events(data: Uint8Array): MidiReport {
  const parsed: [string, Buffer][] = [];
  let cursor = 0;
  while (cursor < data.length) {
    if (cursor + 8 > data.length) throw new Error("truncated MIDI chunk");
    const kind = Buffer.from(data.subarray(cursor, cursor + 4)).toString("ascii");
    const size = new DataView(data.buffer, data.byteOffset + cursor + 4, 4).getUint32(0, false);
    cursor += 8;
    const payload = Buffer.from(data.subarray(cursor, cursor + size));
    cursor += size;
    if (payload.length !== size) throw new Error("truncated MIDI chunk payload");
    parsed.push([kind, payload]);
  }
  if (!parsed.length || parsed[0][0] !== "MThd" || parsed[0][1].length !== 6) throw new Error("invalid MIDI header");
  const form = parsed[0][1].readUInt16BE(0);
  const tracks = parsed[0][1].readUInt16BE(2);
  const division = parsed[0][1].readUInt16BE(4);
  if (![0, 1].includes(form) || tracks === 0 || division & 0x8000) throw new Error("only format 0/1 PPQN MIDI is supported");
  const trackChunks = parsed.slice(1).filter(([kind]) => kind === "MTrk").map(([, payload]) => payload);
  if (trackChunks.length !== tracks) throw new Error("MIDI track count mismatch");
  const result: MidiEvent[] = [];
  for (const [track, payload] of trackChunks.entries()) {
    let cursor = 0, tick = 0, order = 0;
    let running: number | undefined;
    while (cursor < payload.length) {
      let delta: number;
      [delta, cursor] = vlq(payload, cursor);
      tick += delta;
      if (cursor >= payload.length) throw new Error("truncated MIDI event");
      let status = payload[cursor];
      if (status < 0x80) {
        if (running === undefined) throw new Error("running status without channel status");
        status = running;
      } else cursor++;
      const event: MidiEvent = { tick, track, order: order++ };
      if (status === 0xff) {
        if (cursor >= payload.length) throw new Error("truncated meta event");
        const meta = payload[cursor];
        let size: number;
        [size, cursor] = vlq(payload, cursor + 1);
        const value = payload.subarray(cursor, cursor + size);
        cursor += size;
        if (value.length !== size) throw new Error("truncated meta payload");
        Object.assign(event, { type: "meta", meta, data: value.toString("hex") });
        running = undefined;
      } else if (status === 0xf0 || status === 0xf7) {
        let size: number;
        [size, cursor] = vlq(payload, cursor);
        const value = payload.subarray(cursor, cursor + size);
        cursor += size;
        if (value.length !== size) throw new Error("truncated system-exclusive payload");
        Object.assign(event, { type: "sysex", status, data: value.toString("hex") });
        running = undefined;
      } else if (status >= 0x80 && status <= 0xef) {
        running = status;
        const size = [0xc0, 0xd0].includes(status & 0xf0) ? 1 : 2;
        const value = Array.from(payload.subarray(cursor, cursor + size));
        cursor += size;
        if (value.length !== size || value.some((byte) => byte & 0x80)) throw new Error("invalid channel event");
        Object.assign(event, { type: "channel", status, data: value });
      } else throw new Error("unsupported MIDI system event");
      result.push(event);
    }
  }
  result.sort((a, b) => a.tick - b.tick || a.track - b.track || a.order - b.order);
  return { format: form, tracks, ticks_per_quarter: division, events: result };
}

export function make_chunk(kind: Uint8Array, payload: Uint8Array): Buffer {
  const body = Buffer.concat([Buffer.from(kind), Buffer.from(payload)]);
  const head = Buffer.alloc(4), tail = Buffer.alloc(4);
  head.writeUInt32BE(payload.length); tail.writeUInt32BE(crc32(body));
  return Buffer.concat([head, body, tail]);
}

export function make_midi_chunk(kind: Uint8Array, payload: Uint8Array): Buffer {
  const size = Buffer.alloc(4); size.writeUInt32BE(payload.length);
  return Buffer.concat([Buffer.from(kind), size, Buffer.from(payload)]);
}

export async function self_test(): Promise<void> {
  const palette = Buffer.from([0, 0, 0, 248, 0, 0]);
  const rows = Buffer.concat(Array.from({ length: 8 }, () => Buffer.from([0, 0x55])));
  const ihdr = Buffer.alloc(13); ihdr.writeUInt32BE(8, 0); ihdr.writeUInt32BE(8, 4); ihdr.set([1, 3, 0, 0, 0], 8);
  const png = Buffer.concat([Buffer.from("89504e470d0a1a0a", "hex"), make_chunk(Buffer.from("IHDR"), ihdr),
    make_chunk(Buffer.from("PLTE"), palette), make_chunk(Buffer.from("IDAT"), deflateSync(rows)), make_chunk(Buffer.from("IEND"), Buffer.alloc(0))]);
  const [tiles, colors, report] = gba_graphics(png, 4);
  if (!tiles.equals(Buffer.alloc(32, 0x10)) || !colors.equals(Buffer.from([0, 0, 0x1f, 0])) || report.tiles !== 1)
    throw new Error("graphics self-test failed");
  const track = Buffer.from("00c00500903c4060803c0000ff2f00", "hex");
  const midi = Buffer.concat([make_midi_chunk(Buffer.from("MThd"), Buffer.from("000000010060", "hex")), make_midi_chunk(Buffer.from("MTrk"), track)]);
  const midiReport = midi_events(midi);
  if (midiReport.ticks_per_quarter !== 96 || midiReport.events.map((event) => event.tick).join() !== "0,0,96,96")
    throw new Error("MIDI self-test failed");
  const { rgba_image, palette_rgba_image } = await import("./export_asset.ts");
  const rgba = Buffer.from(Array.from({ length: 7 * 5 * 4 }, (_, index) => (index * 29 + 3) & 255));
  const [image] = rgba_image(rgba, 7);
  const unpacked = rgba_png(image);
  if (unpacked[0] !== 7 || unpacked[1] !== 5 || !unpacked[2].equals(rgba)) throw new Error("RGBA self-test failed");
  const rawPalette = Buffer.alloc(256);
  for (let index = 0; index < 128; index++) rawPalette.writeUInt16LE(index * 109 & 0xffff, index * 2);
  const [paletteImage] = palette_rgba_image(rawPalette, 16);
  if (!gba_palette_rgba(paletteImage)[0].equals(rawPalette)) throw new Error("RGBA palette self-test failed");
  console.log("self-test=ok");
}

function option(args: string[], name: string): string {
  const index = args.indexOf(name);
  if (index < 0 || index + 1 >= args.length) throw new Error(`${name} is required`);
  return args[index + 1];
}

function sortedJson(value: Record<string, unknown>): string {
  const sorted = Object.fromEntries(Object.keys(value).sort().map((key) => [key, value[key]]));
  return JSON.stringify(sorted).replace(/:/g, ": ").replace(/,/g, ", ");
}

async function main(args: string[]): Promise<void> {
  if (args.includes("-h") || args.includes("--help")) {
    console.log("usage: import_asset.ts [--self-test] {png,indexed,midi} ...");
    return;
  }
  if (args.includes("--self-test")) {
    await self_test();
    args = args.filter((arg) => arg !== "--self-test");
    if (args.length === 0) return;
  }
  const command = args[0];
  const input = args[1];
  if (!input) throw new Error("an asset command is required");
  if (command === "png") {
    const [tiles, palette, report] = gba_graphics(readFileSync(input), Number(option(args, "--bpp")));
    const tilePath = option(args, "--tiles"), palettePath = option(args, "--palette");
    mkdirSync(dirname(tilePath), { recursive: true }); mkdirSync(dirname(palettePath), { recursive: true });
    writeFileSync(tilePath, tiles); writeFileSync(palettePath, palette); console.log(sortedJson(report));
  } else if (command === "indexed") {
    const output = option(args, args.includes("-o") ? "-o" : "--output");
    const [width, height, pixels, palette] = indexed_png(readFileSync(input));
    mkdirSync(dirname(output), { recursive: true }); writeFileSync(output, Buffer.from(pixels));
    console.log(sortedJson({ width, height, pixels: pixels.length, palette_entries: palette.length }));
  } else if (command === "midi") {
    const output = option(args, args.includes("-o") ? "-o" : "--output");
    const report = midi_events(readFileSync(input));
    mkdirSync(dirname(output), { recursive: true }); writeFileSync(output, canonicalJson(report) + "\n");
    console.log(`tracks=${report.tracks} events=${report.events.length}`);
  } else throw new Error("an asset command is required");
}

if (import.meta.main) await main(process.argv.slice(2));
